consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/FutureMessageBuffer.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.ibft.statemachine;
+
+import tech.pegasys.pantheon.ethereum.p2p.api.Message;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.function.Supplier;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Buffer which holds future IBFT messages.
+ *
+ * <p>This buffer only allows messages to be added which have a chain height greater than current
+ * height and up to chain futureMessagesMaxDistance from the current chain height.
+ *
+ * <p>When the total number of messages is greater futureMessagesLimit then messages are evicted.
+ *
+ * <p>If there is more than one height in the buffer then all messages for the highest chain height
+ * are removed. Otherwise if there is only one height the oldest inserted message is removed.
+ */
+public class FutureMessageBuffer {
+  private final NavigableMap<Long, List<Message>> buffer;
+  private final long futureMessagesMaxDistance;
+  private final long futureMessagesLimit;
+  private final Supplier<Long> currentChainHeight;
+  private long totalMessages;
+
+  public FutureMessageBuffer(
+      final long futureMessagesMaxDistance,
+      final long futureMessagesLimit,
+      final Supplier<Long> currentChainHeight) {
+    this(futureMessagesMaxDistance, futureMessagesLimit, currentChainHeight, new TreeMap<>());
+  }
+
+  @VisibleForTesting
+  FutureMessageBuffer(
+      final long futureMessagesMaxDistance,
+      final long futureMessagesLimit,
+      final Supplier<Long> currentChainHeight,
+      final NavigableMap<Long, List<Message>> buffer) {
+    this.futureMessagesMaxDistance = futureMessagesMaxDistance;
+    this.futureMessagesLimit = futureMessagesLimit;
+    this.currentChainHeight = currentChainHeight;
+    this.buffer = buffer;
+  }
+
+  public void addMessage(final long msgChainHeight, final Message rawMsg) {
+    final long currentHeight = currentChainHeight.get();
+    if (!validMessageHeight(msgChainHeight, currentHeight)) {
+      return;
+    }
+
+    if (totalMessages >= futureMessagesLimit) {
```

Comment:
I had assumed evictMessages() MAY need to loop (i.e. what if you need to remove more than 1 height?) - should this conditional become a loop in evict?

Reply:
I don't think there is any need to remove any more than one height at a time. Each height must contain at least one message so if we remove a height we will have room for the message to be added.